src/mg_decoration.cpp

```diff
@@ -338,16 +338,63 @@ DecoSchematic::DecoSchematic()
 }
 
 
-size_t DecoSchematic::generate(MMVManip *vm, PcgRandom *pr, v3s16 p)
+bool DecoSchematic::canPlaceSchematicDecoration(MMVManip *vm, v3s16 p)
```

Comment:
The first few lines are different, but those can be moved into 'generate' to allow de-duplicating this code, good idea will try this.